lib/cfg.h

```diff
@@ -188,10 +188,36 @@ __cfg_is_config_version_older(GlobalConfig *cfg, gint req)
 
 #define cfg_is_config_version_older(__cfg, __req) \
   ({ \
-    /* check that VERSION_VALUE_LAST_SEMANTIC_CHANGE is set correctly */ G_STATIC_ASSERT((__req) <= VERSION_VALUE_LAST_SEMANTIC_CHANGE); \
+    /* check that VERSION_VALUE_LAST_SEMANTIC_CHANGE is set correctly */ G_STATIC_ASSERT((__req) <= VERSION_VALUE_LAST_SEMANTIC_CHANGE || (__req) >= VERSION_VALUE_4_0); \
     __cfg_is_config_version_older(__cfg, __req); \
   })
 
+/* This function returns TRUE if post-4.0 upgrade warnings should be
+ * displayed or not.  As long as we are using a 3.x version number, these
+ * warnings are suppressed and compatibility mode is enabled.  Once @version
+ * something equal or larger than 3.99, we still enable compatibility but warnings
+ * will NOT be suppressed.
+ *
+ * Once we are ready with all 4.0 changes, we need to do the following:
+ *   1) eliminate all calls to cfg_is_config_to_be_migrated_to_post_4_0()
+ *   2) eliminate the 4.0 related exception from the static assert in
+ *      cfg_is_config_version_older()
+ *   3) set LAST_SEMANTIC_CHANGE to at least 4.0
+ *
+ * With those all upgrade notices start to appear when syslog-ng finds a 3.x
+ * configuration.
+ */
+static inline gboolean
+cfg_is_config_to_be_migrated_to_post_4_0(GlobalConfig *cfg)
+{
+  /* this assert should fail once we release 4.0 */
+  G_STATIC_ASSERT(VERSION_VALUE_CURRENT < VERSION_VALUE_3_LAST);
+
+  if (!cfg)
+    return FALSE;
+  return version_convert_from_user(cfg->user_version) >= VERSION_VALUE_3_LAST;
```

Comment:
The `VERSION_VALUE_CURRENT` can contain higher value then 3.x or even 4.x, and yet it should not work in 4.0 mode.

Reply:
`VERSION_VALUE_3_LAST` is set to 3.99 which I assume will give us enough 3.x versions until we release 4.0

When we do release 4.0 this entire function and all its call sites should be removed.

The static assert in this function should ensure that when we set `VERSION_VALUE_CURRENT` to 4.0 compilation would fail, until at least the assert is removed, but the comment right next to the assert explains that the entire function should be removed at that point.

Does this clarify the intent here? Am I documenting this clearly enough? Or am I missing your point?

Reply:
You are completely missing my point. This code would break our internal building (due to assert), and without the assert it would make our version work in 4.0 mode. It is because you assume everything above 3.99 is 4.0. But in fact it should not. (See https://github.com/balabit-deps/syslog-ng-7.0-core/commit/4aa1246970375664e4eab35bd899a0266f02ed47 )

Reply:
I am not following the PE versioning, but that was the point of version_convert_from_user(cfg->user_version), e.g. to map the users version number to an OSE version.

7.0 could be mapped to VERSION_VALUE_CURRENT